Add backend for fxos feed links

jgmize opened the pull request.

jgmize commented:
TODO:

- [ ] Tests
- [ ] Generate cron job

jgmize commented:
To test locally after pulling down this branch:
```
./manage.py migrate
./manage.py runscript update_firefox_os_feeds
./manage.py switch on firefox-os-index-2015 --create
./manage.py runserver
```
Visit localhost:8000/firefox/os/ and you should see a minimal page with unstyle links created from the blog post feeds that varies based on the locale in the URL.

schalkneethling commented:
Would be awesome if we could get this merged. I can then use the -2015 page from this PR for the work on the new OS page.

jgmize commented:
Rebased, refactored, added tests and cron jobs to templates.
@pmclanahan r?

pmclanahan commented on bedrock/firefox/templates/firefox/os/index-2015.html:
```diff
@@ -0,0 +1,14 @@
+{% add_lang_files "firefox/os/devices" %}
+
+{% block content %}
+{% set feed_links = firefox_os_feed_links(LANG) %}
```
Are we planning on using these links on other pages? If not, why not just add these links in the view?

jgmize replied:
No reason other than it seemed simpler at the time. I'll move this to the view.

jgmize replied:
I forgot: yes, there two other pages that will use this, so it makes more sense as a helper.

pmclanahan replied:
cool.

pmclanahan commented on etc/cron.d/bedrock-stage.tmpl:
```diff
@@ -26,3 +26,5 @@ MAILTO="webops-cron@mozilla.com,cron-bedrock@mozilla.com"
 42 * * * * {{ django_cron }} update_reps_ical > /dev/null 2>&1
 
+# bug 1128587
+38 * * * * {{ django_manage }} runscript update_firefox_os_feeds > /dev/null 2>&1
```
Travis failed because of lack of newline at the end. Need one on prod file as well looks like.

pmclanahan replied:
Maybe we should just add a newline when generating the cron files and stop having these issues.

jgmize replied:
Will fix. Good idea about adding the newline in the generator.

pmclanahan replied:
Filed [bug 1133917](https://bugzilla.mozilla.org/show_bug.cgi?id=1133917) to fix the gen-crons.py script.

pmclanahan commented on scripts/update_firefox_os_feeds.py:
```diff
@@ -0,0 +1,36 @@
+        try:
+            parsed_feed = parse(url)
+        except:
+            print ('Unable to fetch and parse ' + url)
```
Superfluous parens.

jgmize replied:
The parens aren't superfluous, they're forward compatible. I'm in a habit of adding them now from my other project in python 3. I realize we're a long way from running bedrock in python 3, but I don't see any reason not to start including parens in print statements in new code.

pmclanahan replied:
I guess... but they look silly with the space and will be wrong pep8-wise when we do switch right? Be better to be explicit and just import `print_function` from `__future__` don't you think?

jgmize replied:
Good point, I like importing from the future. :)

pmclanahan replied:
![TO THE FUTURE](http://www.gladiatorguards.com/blog_images/blog-62-main.jpg)

pmclanahan commented on scripts/update_firefox_os_feeds.py:
```diff
@@ -0,0 +1,36 @@
+
+def create_or_update_fxos_feed_link(locale, link, title):
+    try:
+        feed_link = FirefoxOSFeedLink.objects.get(
```
Any reason you didn't use `objects.get_or_create()`?

jgmize replied:
Yes, I wanted to handle the title changing in the feed.

pmclanahan replied:
```python
feed_link, created = FirefoxOSFeedLink.objects.get_or_create(locale=locale, link=link, defaults={'title': title})
if not created and title != feed_link.title:
    feed_link.title = title
    feed_link.save()
```

Just a NIT. No need to change since you already typed it all out, just seemed like unnecessary repetition.

jgmize replied:
/me forgot about `defaults`, will update.

pmclanahan commented on bedrock/firefox/helpers.py:
```diff
@@ -0,0 +1,14 @@
+
+@jingo.register.function
+def firefox_os_feed_links(locale):
+    if locale in settings.FIREFOX_OS_FEED_LOCALES:
```
Add some memcache?

jgmize replied:
Nice catch, will do.

jgmize commented:
@pmclanahan I've updated based on your feedback including adding caching (and tests for all the new branches), importing from the future, and newlines for the cron jobs. I also rebased off the current master and squashed, so we should be good to merge in the morning.

@sgarrity you should be able to use the same helper and similar [template code](https://github.com/jgmize/bedrock/blob/firefox-os-feeds/bedrock/firefox/templates/firefox/os/index-2015.html) for the pages that you're working on that will have the same feeds.

pmclanahan commented on bedrock/firefox/tests/test_helpers.py:
```diff
@@ -0,0 +1,53 @@
+        """
+        Should return None without checking cache or db.
+        """
+        eq_(helpers.firefox_os_feed_links('yy'), None)
```
NIT: weird for me to see a test like `x == None` instead of `x is None`. I usually do `self.assertIsNone(x)` or `ok_(x is None)`. This is extremely minor and picky and you shouldn't change it, just wanted to get it out there in case I was missing something.

jgmize replied:
ok, I'll do that next time :)

pmclanahan replied:
Meh... I'm sure it's exactly the same. Just writing what comes to mind :)

pmclanahan commented on bedrock/firefox/tests/test_helpers.py:
```diff
@@ -0,0 +1,53 @@
+        """
+        Should force cache update of first 10 values without cache.get()
+        """
+
```
NIT: extra blank line.

pmclanahan commented on bedrock/settings/base.py:
```diff
@@ -2216,3 +2216,15 @@ def facebook_tab_url_lazy():
+    ('pl', 'https://blog.mozilla.org/press-pl/category/firefox-os/feed/'),
+    ('pt-BR', 'https://blog.mozilla.org/press-br/category/firefox-os/feed/'),
+)
+FIREFOX_OS_FEED_LOCALES = [feed[0] for feed in FIREFOX_OS_FEEDS]
```
This works fine, but why not just have `FIREFOX_OS_FEEDS` be a `dict`?

jgmize replied:
It couldn't be a regular dict because the ordering matters in this implementation: `es-ES` needs to come before `es`.  An `OrderedDict` would work, but would be uglier for no benefit that I can see.

jgmize replied:
scratch that, I was thinking of a previous implementation where the ordering *did* matter, but that's no longer the case. No reason we couldn't change it to a dict now.

pmclanahan replied:
Ah! Glad you clarified. I couldn't see where order mattered. It's fine, was just worried I'd missed something. Stuff like that is always clearer on the r? side :)
